src/polydatum/middleware.py

```diff
@@ -0,0 +1,177 @@
+from typing import Tuple, Any, Dict, Optional, Callable
+
+from polydatum.services import Service
+from polydatum.context import DataAccessContext
+
+
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return isinstance(other, PathSegment) \
+               and self.name == other.name \
+               and self.meta == other.meta
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {".".join([p.name for p in self.path])}')
+
+
+class DeferredDalAttrRequest:
```

Comment:
I find it useful to reference existing design patterns when trying to come up with names.

So this is really a "Command" following the command pattern:

> The command pattern is a Gang of Four design pattern. This is a behavioural pattern as it defines a manner for controlling communication between classes or entities. The command pattern is used to express a request, including the call to be made and all of its required parameters, in a command object. The command may then be executed immediately or held for later use. Often a queue of commands will be created that can be executed in a batch.
> The command object does not contain the functionality that is to be executed, only the information required to perform an action. The functionality is contained within receiver objects. This removes the direct link between the command definitions and the functionality, promoting loose coupling. Neither of these object types is responsible for determining the time of execution of the command. This is controlled using an invoker.

So `DalCommand` is a good enough name. (Not to be confused with `dalcmd` which is a command line dal interface. It would probably be better named `dalcli`). 

Intent

- encapsulate a request in an object
- allows the parameterization of a command for different requests
- pass command as an argument to other functions
- delay execution or put in a queue or use as an event handler

`DalCommand` encapsulates the operation to perform and its parameters. By calling `__call__()` (named `.execute()` in the referenced texts) on the command, it passes the command name and parameters to the Invoker. The Invoker in this case is `DataAccessLayer._call`. The Invoker may enhance the command input further. In this case, the Invoker takes the command input and builds a `DalMethodRequest` with additional context. The Invoker then decides (via middleware) which Receiver to send the request to. The Receiver is the DAL `Service`.

I'm tempted to rename `DalMethodRequest` to `DalCommandRequest`. The fact that it is a method is really inconsequential. (I'd even consider calling it `DalCommandContext` but trying not to confuse it with the other context).

```suggestion
class DalCommand:
    """
    Expresses a request for an operation to be performed.

    It does not directly execute the operation. It passes the requested command path 
    and parameters to the `DataAccessLayer` (DAL). The DAL turns the command into
    a `DalCommandRequest` which also includes the current `DataAccessContext`. The 
    DAL runs the `DalCommandRequest` through middleware which will resolve the 
    command to the specific `Service` that can perform the operation described by the
    command. 

    Middleware may prevent commands from being called by raising exceptions.
    
    Since a `DalCommand` is not bound to a `DataAccessContext`, it can be re-used and
    passed to queues, other functions, or as an event handler to be performed at a 
    later time.     
    """
```

See also:

- http://www.blackwasp.co.uk/Command.aspx
- https://betterprogramming.pub/the-command-design-pattern-2313909122b5
- https://www.tutorialspoint.com/design_pattern/command_pattern.htm
- https://sourcemaking.com/design_patterns/command
- https://refactoring.guru/design-patterns/command
- https://en.wikipedia.org/wiki/Command_pattern

Note: Though I said don't confuse this with `dalcmd`, thinking of the CLI is a way to visualize this. When you type `rm -rf /` on the command line, you aren't actually performing an operation. The CLI is a simplified interface that gives access to operations. `rm -rf /` is a serialized command structure that is easy for humans to type. When the CLI REPL processes the command, it parses it into command names and parameters, then passes it on to the `rm` program as an actual operation to run. 

Note: There is a further nuance here but I'm not sure it's worth the semantics. This `DalCommand` is really a `DalCommandBuilder`. It's not really a "command" until `__call__` is called and parameters are given. A command without parameters isn't really a command yet. So you could consider `DalCommandBuilder` the name and `__call__` produces a `DalCommand` which gets enhanced as a `DalCommandRequest`. 

Further Note: `PathSegment` begins to feel off here. It's not really a "path" that we normally think of. `CommandName`? Nah...

Note Note Note: We call it "method middleware" and "blah blah dal method". Maybe "command middleware" and "command" is more appropriate terminology.
 
- [ ] Rename to `DalCommand`